modules/xfd.ts

```diff
@@ -0,0 +1,2218 @@
+	}
+
+	// Not overridden for any venue
+	getNotifyEditSummary(): string {
+		return 'Notification: [[' + this.params.discussionpage + '|listing]] of [[:' +
+		Morebits.pageNameNorm + ']] at [[WP:' + this.getFieldsetLabel() + ']].';
+	}
+
+	notifyCreator(): JQuery.Promise<void> {
+		if (!this.params.notifycreator) {
+			this.params.intialContrib = null;
+			return $.Deferred().resolve();
+		}
+		return this.notifyTalkPage(this.params.initialContrib);
+	}
+
+	// Should be called after notifyTalkPage() which may unset this.params.intialContrib
+	addToLog() {
+		let params = this.params;
+
+		if (!Twinkle.getPref('logXfdNominations') ||
+			Twinkle.getPref('noLogOnXfdNomination').indexOf(params.venue) !== -1) {
+			return $.Deferred().resolve();
+		}
+
+		var usl = new Morebits.userspaceLogger(Twinkle.getPref('xfdLogPageName'));// , 'Adding entry to userspace log');
+
+		usl.initialText =
+			"This is a log of all [[WP:XFD|deletion discussion]] nominations made by this user using [[WP:TW|Twinkle]]'s XfD module.\n\n" +
+			'If you no longer wish to keep this log, you can turn it off using the [[Wikipedia:Twinkle/Preferences|preferences panel]], and ' +
+			'nominate this page for speedy deletion under [[WP:CSD#U1|CSD U1]].' +
+			(Morebits.userIsSysop ? '\n\nThis log does not track XfD-related deletions made using Twinkle.' : '');
+
+		usl.changeTags = Twinkle.changeTags;
+		return usl.log(this.getUserspaceLoggingText(), this.getUserspaceLoggingEditSummary());
+	}
+
+	getUserspaceLoggingEditSummary() {
+		return 'Logging ' + utils.toTLACase(this.params.venue) + ' nomination of [[:' + Morebits.pageNameNorm + ']].';
+	}
+
+	getUserspaceLoggingText(): string {
+		let params = this.params;
+
+		// If a logged file is deleted but exists on commons, the wikilink will be blue, so provide a link to the log
+		var fileLogLink = mw.config.get('wgNamespaceNumber') === 6 ? ' ([{{fullurl:Special:Log|page=' + mw.util.wikiUrlencode(mw.config.get('wgPageName')) + '}} log])' : '';
+		// CFD/S and RM don't have canonical links
+		var nominatedLink = params.discussionpage ? '[[' + params.discussionpage + '|nominated]]' : 'nominated';
+
+		var appendText = '# [[:' + Morebits.pageNameNorm + ']]:' + fileLogLink + ' ' + nominatedLink + ' at [[WP:' + params.venue.toUpperCase() + '|' + utils.toTLACase(params.venue) + ']]';
+
+		appendText += this.getUserspaceLoggingExtraInfo();
+
+		if (params.initialContrib && params.notifycreator) {
+			appendText += '; notified {{user|1=' + params.initialContrib + '}}';
+		}
+		appendText += ' ~~~~~';
+		if (params.reason) {
+			appendText += "\n#* '''Reason''': " + Morebits.string.formatReasonForLog(params.reason);
+		}
+		return appendText;
+	}
+
+	getUserspaceLoggingExtraInfo() {
+		return '';
+	}
+
+}
+
+class Afd extends XfdMode {
+	static venueCode = 'afd';
+	static venueLabel = 'AfD (Articles for deletion)';
+
+	static isDefaultChoice() {
+		return mw.config.get('wgNamespaceNumber') === 0;
+	}
+
+	getFieldsetLabel() {
```

Comment:
why not have the field set label as a static value, like venueCode and venueLabel?

Reply:
It has to do with how it's invoked – [on line 162](https://github.com/siddharthvp/twinkle/blob/8e12b1070f3a104fc37ee4e285c45e3933fb177f/modules/xfd.ts#L162). Here we call `getFieldsetLabel` on `this` without actually knowing what class `this` is an object of – it could be any of `Afd`, `Tfd`, etc.

Reply:
I take your point that per the spirit of OOP, the method should be static, as the value is the same for all objects constructed from the class. In a truly object-oriented language like Java, this could've been made static since `cls.staticMethod()` and `obj.staticMethod()` both work (in Java), but in JS/TS, static methods can unfortunately only be called on the class itself, not on the objects.